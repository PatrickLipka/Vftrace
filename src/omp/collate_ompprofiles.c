#include <stdlib.h>

#include <string.h>

#ifdef _MPI
#include <mpi.h>
#endif

#include "self_profile.h"
#include "ompprofiling_types.h"
#include "collated_stack_types.h"
#include "stack_types.h"

static void vftr_collate_ompprofiles_root_self(collated_stacktree_t *collstacktree_ptr,
                                               stacktree_t *stacktree_ptr) {
   SELF_PROFILE_START_FUNCTION;
   for (int istack=0; istack<stacktree_ptr->nstacks; istack++) {
      stack_t *stack = stacktree_ptr->stacks+istack;
      int icollstack = stack->gid;

      collated_stack_t *collstack = collstacktree_ptr->stacks+icollstack;
      ompprofile_t *collompprof = &(collstack->profile.ompprof);

      collompprof->overhead_nsec = 0ll;

      for (int iprof=0; iprof<stack->profiling.nprofiles; iprof++) {
         ompprofile_t *ompprof = &(stack->profiling.profiles[iprof].ompprof);
   
         collompprof->overhead_nsec += ompprof->overhead_nsec;
      }
   }
   SELF_PROFILE_END_FUNCTION;
}

#ifdef _MPI
static void vftr_collate_ompprofiles_on_root(collated_stacktree_t *collstacktree_ptr,
                                             stacktree_t *stacktree_ptr,
                                             int myrank, int nranks,
                                             int *nremote_profiles) {
   SELF_PROFILE_START_FUNCTION;
   // define datatypes required for collating ompprofiles
   typedef struct {
      int gid;
      long long overhead_nsec;
   } ompprofile_transfer_t;

   int nblocks = 2;
   const int blocklengths[] = {1,1};
   const MPI_Aint displacements[] = {0, sizeof(int)};
   const MPI_Datatype types[] = {MPI_INT, MPI_LONG_LONG_INT};
   MPI_Datatype ompprofile_transfer_mpi_t;
   PMPI_Type_create_struct(nblocks, blocklengths,
                           displacements, types,
                           &ompprofile_transfer_mpi_t);
   PMPI_Type_commit(&ompprofile_transfer_mpi_t);

   if (myrank > 0) {
      // every rank fills their sendbuffer
      int nprofiles = stacktree_ptr->nstacks;
      ompprofile_transfer_t *sendbuf = (ompprofile_transfer_t*)
         malloc(nprofiles*sizeof(ompprofile_transfer_t));
      for (int istack=0; istack<nprofiles; istack++) {
         sendbuf[istack].gid = 0;
         sendbuf[istack].total_time_nsec = 0ll;
      }
      for (int istack=0; istack<nprofiles; istack++) {
         stack_t *mystack = stacktree_ptr->stacks+istack;
         sendbuf[istack].gid = mystack->gid;
         // need to go over the calling profiles threadwise
         for (int iprof=0; iprof<mystack->profiling.nprofiles; iprof++) {
            profile_t *myprof = mystack->profiling.profiles+iprof;
            ompprofile_t ompprof = myprof->ompprof;
            sendbuf[istack].overhead_nsec += ompprof.overhead_nsec;
         }
      }
      PMPI_Send(sendbuf, nprofiles,
                ompprofile_transfer_mpi_t,
                0, myrank,
                MPI_COMM_WORLD);
      free(sendbuf);
   } else {
      int maxprofiles = 0;
      for (int irank=1; irank<nranks; irank++) {
         maxprofiles = nremote_profiles[irank] > maxprofiles ? 
                       nremote_profiles[irank] :
                       maxprofiles;
      }
      ompprofile_transfer_t *recvbuf = (ompprofile_transfer_t*)
         malloc(maxprofiles*sizeof(ompprofile_transfer_t));
      memset(recvbuf, 0, maxprofiles*sizeof(ompprofile_transfer_t));
      for (int irank=1; irank<nranks; irank++) {
         int nprofiles = nremote_profiles[irank];
         MPI_Status status;
         PMPI_Recv(recvbuf, nprofiles,
                   ompprofile_transfer_mpi_t,
                   irank, irank,
                   MPI_COMM_WORLD,
                   &status);
         for (int iprof=0; iprof<nprofiles; iprof++) {
            int gid = recvbuf[iprof].gid;
            collated_stack_t *collstack = collstacktree_ptr->stacks+gid;
            ompprofile_t *collompprof = &(collstack->profile.ompprof);
     
            collompprof->overhead_nsec += recvbuf[iprof].overhead_nsec;
         }
      }
      free(recvbuf);
   }

   PMPI_Type_free(&ompprofile_transfer_mpi_t);
   SELF_PROFILE_END_FUNCTION;
}
#endif

void vftr_collate_ompprofiles(collated_stacktree_t *collstacktree_ptr,
                              stacktree_t *stacktree_ptr,
                              int myrank, int nranks,
                              int *nremote_profiles) {
   SELF_PROFILE_START_FUNCTION;
   vftr_collate_ompprofiles_root_self(collstacktree_ptr, stacktree_ptr);
#ifdef _MPI
   int mpi_initialized;
   PMPI_Initialized(&mpi_initialized);
   if (mpi_initialized) {
      vftr_collate_ompprofiles_on_root(collstacktree_ptr, stacktree_ptr,
                                       myrank, nranks, nremote_profiles);
   }
#else
   (void) myrank;
   (void) nranks;
   (void) nremote_profiles;
#endif
   SELF_PROFILE_END_FUNCTION;
}
