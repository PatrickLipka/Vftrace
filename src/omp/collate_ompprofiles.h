#ifndef COLLATE_OMPPROFILES_H
#define COLLATE_OMPPROFILES_H

#include <stdlib.h>

#include "stack_types.h"
#include "collated_stack_types.h"

void vftr_collate_ompprofiles(collated_stacktree_t *collstacktree_ptr,
                              stacktree_t *stacktree_ptr,
                              int myrank, int nranks,
                              int *nremote_profiles);

#endif
