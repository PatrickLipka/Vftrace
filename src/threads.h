#ifndef THREADS_H
#define THREADS_H

#include <stdbool.h>

#include "thread_types.h"

threadtree_t vftr_new_threadtree();

void vftr_threadtree_free(threadtree_t *threadtree_ptr);

thread_t *vftr_get_my_thread(threadtree_t threadtree);

#endif
