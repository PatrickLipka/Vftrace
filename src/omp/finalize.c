#include <omp.h>
#include <omp-tools.h>

void ompt_finalize(ompt_data_t *tool_data) {
   (void) tool_data;
}
