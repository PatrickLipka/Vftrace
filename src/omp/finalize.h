#ifndef FINALIZE_H
#define FINALIZE_H

#include <omp.h>
#include <omp-tools.h>

void ompt_finalize(ompt_data_t *tool_data);

//extern void (*ompt_finalize_ptr)(ompt_data_t*);

#endif
