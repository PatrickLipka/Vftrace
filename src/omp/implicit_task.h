#ifndef IMPLICIT_TASK_H
#define IMPLICIT_TASK_H

#include <omp.h>
#include <omp-tools.h>

static void vftr_ompt_callback_implicit_task(ompt_scope_endpoint_t endpoint,
                                             ompt_data_t *parallel_data,
                                             ompt_data_t *task_data,
                                             unsigned int actual_parallelism,
                                             unsigned int index,
                                             int flags);

void vftr_register_ompt_callback_implicit_task(ompt_set_callback_t ompt_set_callback);

#endif
