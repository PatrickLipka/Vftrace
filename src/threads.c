#include <stdlib.h>
#include <stdbool.h>

#ifdef _OMP
#include <omp.h>
#endif

#include "realloc_consts.h"
#include "stacks.h"
#include "threads.h"
#include "threadstacks.h"

void vftr_threadtree_realloc(threadtree_t *threadtree_ptr) {
   threadtree_t threadtree = *threadtree_ptr;
   while (threadtree.nthreads > threadtree.maxthreads) {
      int maxthreads = threadtree.maxthreads*vftr_realloc_rate+vftr_realloc_add;
      threadtree.threads = (thread_t*)
         realloc(threadtree.threads, maxthreads*sizeof(thread_t));
      threadtree.maxthreads = maxthreads;
   }
   *threadtree_ptr = threadtree;
}

void vftr_thread_subthreads_realloc(thread_t *thread_ptr) {
   thread_t thread = *thread_ptr;
   while (thread.nsubthreads > thread.maxsubthreads) {
      int maxsubthreads = thread.maxsubthreads*vftr_realloc_rate+vftr_realloc_add;
      thread.subthreads = (int*)
         realloc(thread.subthreads, maxsubthreads*sizeof(int));
      thread.maxsubthreads = maxsubthreads;
   }
   *thread_ptr = thread;
}

int vftr_new_thread(int parent_thread_id,
                    threadtree_t *threadtree_ptr) {
   thread_t thread;
   thread.threadID = threadtree_ptr->nthreads;
   threadtree_ptr->nthreads++;
   vftr_threadtree_realloc(threadtree_ptr);
   thread_t *parent_thread_ptr = threadtree_ptr->threads+parent_thread_id;
   thread.level = parent_thread_ptr->level+1;
   thread.thread_num = parent_thread_ptr->nsubthreads;
   thread.master = parent_thread_ptr->master && thread.thread_num == 0;
   thread.stacklist = vftr_new_threadstacklist(-1);
   thread.parent_thread = parent_thread_ptr->threadID;
   thread.maxsubthreads = 0;
   thread.nsubthreads = 0;
   thread.subthreads = NULL;
   // add thread to threadtree
   threadtree_ptr->threads[thread.threadID] = thread;
   // add it to the subthread list of the parent thread
   parent_thread_ptr->nsubthreads++;

   vftr_thread_subthreads_realloc(parent_thread_ptr);
//printf("parent subthreads = %d\n", parent_thread_ptr->nsubthreads);
   parent_thread_ptr->subthreads[thread.thread_num] = thread.threadID;

   return thread.threadID;
}

thread_t vftr_new_masterthread() {
   thread_t thread;
   thread.level = 0;
   thread.thread_num = 0;
   thread.master = true;
   thread.parent_thread = -1;
   thread.threadID = 0;
   thread.stacklist = vftr_new_threadstacklist(0);
   thread.maxsubthreads = 0;
   thread.nsubthreads = 0;
   thread.subthreads = NULL;
   return thread;
}

threadtree_t vftr_new_threadtree() {
   threadtree_t threadtree;
   threadtree.nthreads = 1;
   threadtree.maxthreads = 1;
   threadtree.threads = (thread_t*) malloc(sizeof(thread_t));
   threadtree.threads[0] = vftr_new_masterthread();
   return threadtree;
}

int vftr_get_thread_level() {
#ifdef _OMP
   return omp_get_level();
#else
   return 0;
#endif
}

int vftr_get_thread_num() {
#ifdef _OMP
   return omp_get_thread_num();
#else
   return 0;
#endif
}

int vftr_get_ancestor_thread_num(int level) {
#ifdef _OMP
   return omp_get_ancestor_thread_num(level);
#else
   (void) level;
   return 0;
#endif
}

thread_t *vftr_get_my_thread(threadtree_t *threadtree_ptr) {
   int level = vftr_get_thread_level();
   thread_t *my_thread = threadtree_ptr->threads;
   // navigate through the thread tree until my thread is found
   for (int ilevel=1; ilevel<level; ilevel++) {
      int thread_num = vftr_get_ancestor_thread_num(ilevel);
      int threadID = my_thread->nsubthreads;
      // if the thread does not exist yet, add and null it.
      while (thread_num > my_thread->nsubthreads) {
         threadID = vftr_new_thread(my_thread->threadID,
                                    threadtree_ptr);
      }
      my_thread = threadtree_ptr->threads+threadID;
   }
   return my_thread;
}

void vftr_thread_free(thread_t *threads_ptr, int threadID) {
   thread_t thread = threads_ptr[threadID];
   vftr_threadstacklist_free(&(thread.stacklist));
   if (thread.nsubthreads > 0) {
      for (int ithread=0; ithread<thread.nsubthreads; ithread++) {
         vftr_thread_free(threads_ptr, thread.subthreads[ithread]);
      }
      free(thread.subthreads);
      thread.subthreads = NULL;
   }
   threads_ptr[threadID] = thread;
}

void vftr_threadtree_free(threadtree_t *threadtree_ptr) {
   threadtree_t threadtree = *threadtree_ptr;
   if (threadtree.nthreads > 0) {
      vftr_thread_free(threadtree.threads, 0);
      free(threadtree.threads);
      threadtree.threads = NULL;
      threadtree.nthreads = 0;
      threadtree.maxthreads = 0;
   }
   *threadtree_ptr = threadtree;
}

void vftr_print_thread(FILE *fp, threadtree_t threadtree, int threadid) {
   thread_t thread = threadtree.threads[threadid];
   // first print the indentation
   for (int ilevel=0; ilevel<thread.level; ilevel++) {
      fprintf(fp, "  ");
   }
   fprintf(fp, "%d%s: ", thread.thread_num,
           thread.master ? "m" : "");
   vftr_print_threadstack(fp, thread.stacklist);
   fprintf(fp, "\n");
   for (int ithread=0; ithread<thread.nsubthreads; ithread++) {
      vftr_print_thread(fp, threadtree, thread.subthreads[ithread]);
   }
}

void vftr_print_threadtree(FILE *fp, threadtree_t threadtree) {
   fprintf(fp, "Threadtree:\n");
   vftr_print_thread(fp, threadtree, 0);
}
