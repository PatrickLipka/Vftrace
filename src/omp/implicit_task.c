#include <stdio.h>

#include <omp.h>
#include <omp-tools.h>

#include "threads.h"
#include "vftrace_state.h"

static void vftr_ompt_callback_implicit_task(ompt_scope_endpoint_t endpoint,
                                             ompt_data_t *parallel_data,
                                             ompt_data_t *task_data,
                                             unsigned int actual_parallelism,
                                             unsigned int index,
                                             int flags) {
//   switch (endpoint) {
//      case ompt_scope_begin:
//         //thread_t *my_thread = vftr_get_my_thread(&vftrace.process.threadtree);
//
//         fprintf(stderr, "implicit_task begin:  t=%d l=%d\n", omp_get_thread_num(), omp_get_level());
//         break;
//      case ompt_scope_end:
//         fprintf(stderr, "implicit_task end:    t=%d l=%d\n", omp_get_thread_num(), omp_get_level());
//         break;
//      case ompt_scope_beginend:
//         printf("implicit_task begend: t=%d\n", omp_get_thread_num());
//         break;
//      default:
//         break;
//   }
}

void vftr_register_ompt_callback_implicit_task(ompt_set_callback_t ompt_set_callback) {
   ompt_callback_implicit_task_t f_ompt_callback_implicit_task =
      &vftr_ompt_callback_implicit_task;
   int ompt_set =  ompt_set_callback(ompt_callback_implicit_task,
                                     (ompt_callback_t)f_ompt_callback_implicit_task);
   if (ompt_set == ompt_set_never) {
      fprintf(stderr, "0: Could not register callback \"ompt_callback_implicit_task\"\n");
   }
}
