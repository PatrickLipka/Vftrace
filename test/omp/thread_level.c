#include <stdio.h>
#include <stdbool.h>

#include <omp.h>

#include "threads.h"

int main(int argc, char **argv) {
   int level = vftr_get_thread_level();
   int thread = vftr_get_thread_num();
   printf("Thread %d at level %d\n", thread, level);
   #pragma omp parallel num_threads(2) private(level,thread)
   {
      level = vftr_get_thread_level();
      thread = vftr_get_thread_num();
      printf("Thread %d at level %d\n", thread, level);
      if (thread == 1) {
         #pragma omp parallel num_threads(2) private(level,thread)
         {
            level = vftr_get_thread_level();
            thread = vftr_get_thread_num();
            printf("Thread %d at level %d\n", thread, level);
         }
      }
   }

   return 0;
}
