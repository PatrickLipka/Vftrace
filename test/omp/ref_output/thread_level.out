Thread 0 at level 0
Thread 0 at level 1
Thread 1 at level 1
Thread 0 at level 2
Thread 1 at level 2
